### ISSUE TITLE
Error handling in FileSystemProvider.pathIteratorForPublisher()

### ISSUE THREAD

**Issue by stefanofornari** [awslabs/aws-java-nio-spi-for-s3#189]
pathIteratorForPublisher() makes use of Flowable to retrieve paginated results from ListObjectsV2Request (through pathIteratorForPublisher() ):

https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/ef8f0ff5b56b0a2123eff8aaa4aa5b4949c2cfe7/src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java#L431-L466

If the bucket does not exist, or access to it is denied, the underlying S3 client throws an exception but the error condition is not handled (see an example of trace below).

We should have a way to handle or propagate the error consistently with the other NIO methods. In trying to address the issue, before everything else, I want to have a test that reproduces the error... and I am struggling to build a proper mock. In essence, I need to mock a client so that listObjectsV2() throws an sdk error. Something like:

```
when(mockClient.listObjectsV2(any(ListObjectsV2Request.class))).thenReturn(
    CompletableFuture.supplyAsync(
        () -> ListObjectsV2Response.builder().sdkHttpResponse(
            SdkHttpResponse.builder().statusCode(403).build()
        ).build()
    )
);
```

The code above does not work, @markjschreiber can you please help me out?

Exception thrown: 

```
Exception in thread "main" java.util.concurrent.CompletionException: software.amazon.awssdk.services.s3.model.NoSuchBucketException: The specified bucket does not exist (Service: S3, Status Code: 404, Request ID: 1788C59F052E45EF)
        at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:65)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncExecutionFailureExceptionReportingStage.lambda$execute$0(AsyncExecutionFailureExceptionReportingStage.java:51)
        at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
        at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
        at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:79)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeAttemptExecute(AsyncRetryableStage.java:103)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage.java:184)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.lambda$attemptExecute$1(AsyncRetryableStage.java:170)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$null$0(MakeAsyncHttpRequestStage.java:105)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.completeResponseFuture(MakeAsyncHttpRequestStage.java:238)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$3(MakeAsyncHttpRequestStage.java:163)
        at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
        at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
        at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: software.amazon.awssdk.services.s3.model.NoSuchBucketException: The specified bucket does not exist (Service: S3, Status Code: 404, Request ID: 1788C59F052E45EF)
        at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleErrorResponse(AwsXmlPredicatedResponseHandler.java:156)
        at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleResponse(AwsXmlPredicatedResponseHandler.java:108)
        at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:85)
        at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:43)
        at software.amazon.awssdk.core.internal.handler.BaseClientHandler.lambda$successTransformationResponseHandler$7(BaseClientHandler.java:270)
        at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:89)
        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:132)
        at software.amazon.awssdk.utils.async.SimplePublisher.doProcessQueue(SimplePublisher.java:275)
        at software.amazon.awssdk.utils.async.SimplePublisher.processEventQueue(SimplePublisher.java:224)
        at software.amazon.awssdk.utils.async.SimplePublisher.complete(SimplePublisher.java:157)
        at java.base/java.util.concurrent.CompletableFuture.uniRunNow(CompletableFuture.java:819)
        at java.base/java.util.concurrent.CompletableFuture.uniRunStage(CompletableFuture.java:803)
        at java.base/java.util.concurrent.CompletableFuture.thenRun(CompletableFuture.java:2227)
        at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onErrorResponseComplete(S3CrtResponseHandlerAdapter.java:143)
        at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.handleError(S3CrtResponseHandlerAdapter.java:132)
        at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onFinished(S3CrtResponseHandlerAdapter.java:101)
        at software.amazon.awssdk.crt.s3.S3MetaRequestResponseHandlerNativeAdapter.onFinished(S3MetaRequestResponseHandlerNativeAdapter.java:24)

```

**Comment by markjschreiber** [comment]
To get a mock to throw an exception you can use `doThrow` or `thenThrow`. Take a look at https://www.baeldung.com/mockito-exceptions for an example.

**Comment by stefanofornari** [comment]
Not really, I'd like to make it return 403 or 404 as in the snippet and like it is done in checkAccess()

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by markjschreiber** [comment]
Do you know if the mock client is actually being called? It might be worth running the test with a debugger if you can to step into it.

This error `software.amazon.awssdk.services.s3.model.NoSuchBucketException: The specified bucket does not exist (Service: S3, Status Code: 404, Request ID: 1788C59F052E45EF)` seems to indicate a call is being made to a real client (somewhere) and this interaction might also need to be mocked.

**Comment by stefanofornari** [comment]
Hi @markjschreiber , the error is due to a real access. In fact what I am asking is to reproduce the same error with a proper Mockj, think that I have not figured out how to do.

**Comment by guicamest** [comment]
Is this closed by https://github.com/awslabs/aws-java-nio-spi-for-s3/pull/303/commits/c48db28650abae5ce6dfeff88336f34942ad5f45 ?

**Comment by stefanofornari** [comment]
I do not think so, I want it done at build time, not as an integration test.

**Comment by guicamest** [comment]
We can make the `build` task depend on the integration tests in that case, what do you think?

**Comment by stefanofornari** [comment]
Not at all. Integration tests should be something additional and optional to capture issues with real systems. During development the lighter, the better.

**Comment by markjschreiber** [comment]
After some investigation it seems that when there is an exception due to permissions or bucket not existing you don't get a `ListObjectsV2Response` with an error status code rather you get an `SDKException` with the corresponding status code. So in the mock to simulate various error types you would have it throw an exception in response to the request.

**Comment by stefanofornari** [comment]
Exactly, how can I build that mock? I have tried but I could not get it work ... :(

**Comment by markjschreiber** [comment]
The general pattern would be `when(myMock.doSomething()).thenThrow(new MyException());`

Where `MyException` would be an appropriate `SdkException` with the status code you want. Then you assert that it is handled as expected.

**Comment by markjschreiber** [comment]
This should be fixed now in #404. `RuntimeExceptions` and `SdkExceptions` are perpetuated up as `IOExceptions`
